ifsbench/arch.py

```diff
@@ -9,467 +9,140 @@
 Architecture specifications
 """
 from abc import ABC, abstractmethod
-import os
+from dataclasses import dataclass, field
+from typing import List
 
-from ifsbench.job import CpuConfiguration, CpuBinding, Job
-from ifsbench.launcher import Launcher, MpirunLauncher, SrunLauncher, AprunLauncher
-from ifsbench.util import as_tuple, execute
+from ifsbench.env import EnvHandler
+from ifsbench.launcher.launcher import Launcher
 
+__all__ = ['Arch', 'DefaultArch']
 
-__all__ = ['Arch', 'Workstation', 'XC40Cray', 'XC40Intel', 'AtosAaIntel',
-           'LumiC', 'LumiG', 'arch_registry']
-
-
-class Arch(ABC):
+@dataclass
+class ArchResult:
     """
-    Machine and compiler architecture on which to run the IFS
-
-    This provides compiler and environment assumptions for MPI-parallel runs.
+    Holds results of an :meth:`Arch.process` run.
 
-    An implementation of this class must be provided for each target system.
-    It must declare the CPU configuration in :attr:`cpu_config` and the
-    launcher in :attr:`launcher` that is used for MPI-parallel execution.
-
-    For multiple toolchains/runtime environments on a single physical system
-    it can be beneficial to create an intermediate class describing common
-    properties (such as :attr:`cpu_config`) and derived classes to specify
-    the bespoke configuration.
+    Attributes:
+    job: Job
+        The updated job after the architecture processing.
+    env_handlers: list[EnvHandler]
+        Additional EnvHandler objects that set architecture-specific environment
+        flags.
+    default_launcher: Launcher
+        The default launcher that is used on this system.
+    default_launcher_flags : list[str]
+        Additional launcher flags that should be added to launcher invocations.
     """
+    job = None
+    env_handlers : List[EnvHandler] = field(default_factory=list)
+    default_launcher : Launcher = None
+    default_launcher_flags : List[str] = field(default_factory=list)
```

Comment:
Docstrings won't render right, here.

```suggestion
    """
    #: The updated job after the architecture processing.
    job = None
    
    #: Additional EnvHandler objects that set architecture-specific environment flags
    env_handlers : List[EnvHandler] = field(default_factory=list)
    
    #: The default launcher that is used on this system.
    default_launcher : Launcher = None
    
    #: Additional launcher flags that should be added to launcher invocations.
    default_launcher_flags : List[str] = field(default_factory=list)
```